From: SJ Park <sj@kernel.org>
Date: Mon, 24 Aug 2026 21:36:33 -0700
Subject: [PATCH] mm/damon/core: introduce damos_quota_goal->complement

Introduce damos_quota_goal->complement for using complented value of the
goal target metric.  Add the field and update the goal commit logic to
do update the field of destination goal with the value of the given
source.

Signed-off-by: SJ Park <sj@kernel.org>
---
 include/linux/damon.h |  2 ++
 mm/damon/core.c       | 10 ++++++++++
 2 files changed, 12 insertions(+)

diff --git a/include/linux/damon.h b/include/linux/damon.h
index 836353c4ab9aa..7659af9b218bf 100644
--- a/include/linux/damon.h
+++ b/include/linux/damon.h
@@ -175,6 +175,7 @@ enum damos_quota_goal_metric {
 /**
  * struct damos_quota_goal - DAMOS scheme quota auto-tuning goal.
  * @metric:		Metric to be used for representing the goal.
+ * @complement:		Use the complement of the metric.
  * @target_value:	Target value of @metric to achieve with the tuning.
  * @current_value:	Current value of @metric.
  * @nid:		Node id.
@@ -197,6 +198,7 @@ enum damos_quota_goal_metric {
  */
 struct damos_quota_goal {
 	enum damos_quota_goal_metric metric;
+	bool complement;
 	unsigned long target_value;
 	unsigned long current_value;
 	/* metric-dependent fields */
diff --git a/mm/damon/core.c b/mm/damon/core.c
index 4687b909d42c9..beb55e81d334c 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -1224,6 +1224,7 @@ static int damos_commit_quota_goal(
 	if (!src->target_value)
 		return  -EINVAL;
 	dst->metric = src->metric;
+	dst->complement = src->complement;
 	dst->target_value = src->target_value;
 	if (dst->metric == DAMOS_QUOTA_USER_INPUT)
 		dst->current_value = src->current_value;
@@ -3239,6 +3240,15 @@ static void damos_set_quota_goal_current_value(struct damon_ctx *c,
 	default:
 		break;
 	}
+	switch (goal->metric) {
+	case DAMOS_QUOTA_USER_INPUT:
+	case DAMOS_QUOTA_SOME_MEM_PSI_US:
+		return;
+	default:
+		break;
+	}
+	if (goal->complement)
+		goal->current_value = 10000 - goal->current_value;
 }
 
 /* Return the highest score since it makes schemes least aggressive */
-- 
2.47.3

